### PR TITLE
feat: implement new functions

### PR DESCRIPTION
**Pull request by sillyElvy**
## About

This PR:
- Adds new functions:
  - Channel:
    - pin_message()
    - unpin_message()
    - publish_message()
    - get_pinned_messages()
  - Message:
    - pin()
    - unpin()
    - publish()
  - Guild:
    - get_all_channels()
    - get_all_roles()
    - modify_role_position()
    - get_bans()
  - Role:
    - modify_position()

- Adds missing `return` and `rtype` to docstrings 

- Allows component sending in `Member.send()`

- fixes component sending

## Checklist

- [x] I've ran `pre-commit` to format and lint the change(s) made.
- [x] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [ ] I've made this pull request for/as: (check all that apply)
  - [x] Documentation
  - [ ] Breaking change
  - [x] New feature/enhancement
  - [x] Bugfix

### CODE REVIEW

File: interactions/api/models/channel.py [hunk 1]
```diff
@@ -332,6 +332,61 @@ async def add_member(
...
+
+    async def publish_message(
+        self,
+        message_id: int,
+    ):
+        """Publishes (API calls it crossposts) a message in the channel to any that is followed by.
+
+        :param message_id: The id of the message to publish
+        :type message_id: int
+        :return: message object
```
Comment (i0bs) [inline interactions/api/models/channel.py]:
```suggestion
        :return: The message published
```

File: interactions/api/models/channel.py [hunk 2]
```diff
@@ -332,6 +332,61 @@ async def add_member(
...
+        :rtype: Message
+        """
+        from .message import Message
+
+        res = await self._client.publish_message(
+            channel_id=int(self.id), message_id=int(message_id)
+        )
+        return Message(**res, _client=self._client)
+
+    async def get_pinned_messages(self):
```
Comment (i0bs) [inline interactions/api/models/channel.py:471]:
```suggestion
    async def get_pinned_messages(self) -> List[Message]:
```

File: interactions/api/models/channel.py [hunk 3]
```diff
@@ -332,6 +332,61 @@ async def add_member(
...
+        """
+        Get all pinned messages from the channel.
+        :return: A list of pinned message objects.
```
Comment (i0bs) [inline interactions/api/models/channel.py:474]:
```suggestion
        :return: A list of pinned message objects.
        :rtype: List[Message]
```

File: interactions/api/models/guild.py [hunk 4]
```diff
@@ -1031,6 +1042,65 @@ async def delete_scheduled_event(self, event_id: int) -> None:
...
+        """
+        res = self._client.get_all_channels(int(self.id))
+        channels = [Channel(**channel, _client=self._client) for channel in res]
+        return channels
+
+    async def get_all_roles(self) -> List[Role]:
+        """
+        Gets all roles of the guild as list
+
+        :rtype: List[Role]
```
Comment (i0bs) [inline interactions/api/models/guild.py:1061]:
```suggestion
        :return: The roles of the guild.
        :rtype: List[Role]
```

File: interactions/api/models/guild.py [hunk 5]
```diff
@@ -1031,6 +1042,65 @@ async def delete_scheduled_event(self, event_id: int) -> None:
             guild_scheduled_event_id=Snowflake(event_id),
         )
 
+    async def get_all_channels(self) -> List[Channel]:
+        """
+        Gets all channels of the guild as list
+
+        :rtype: List[Channel]
```
Comment (i0bs) [inline interactions/api/models/guild.py:1050]:
```suggestion
        :return: The channels of the guild.
        :rtype: List[Channel]
```

File: interactions/api/models/guild.py [hunk 6]
```diff
@@ -1031,6 +1042,65 @@ async def delete_scheduled_event(self, event_id: int) -> None:
...
+        Gets a list of banned users
+
+        :return: List of banned users with reasons
+        :rtype: List[dict]
+        """
+
+        res = await self._client.get_guild_bans(int(self.id))
+        for ban in res:
+            ban["user"] = User(**ban["user"])
+        return res
```
Comment (i0bs) [inline interactions/api/models/guild.py:1104]:
```suggestion
        return ban["user"]
```

This should be returning them in the converted object.

File: interactions/api/models/guild.py [hunk 7]
```diff
@@ -1031,6 +1042,65 @@ async def delete_scheduled_event(self, event_id: int) -> None:
...
+        )
+        roles = [Role(**role, _client=self._client) for role in res]
+        return roles
+
+    async def get_bans(self) -> List[dict]:
+        """
+        Gets a list of banned users
+
+        :return: List of banned users with reasons
+        :rtype: List[dict]
```
Comment (i0bs) [inline interactions/api/models/guild.py:1098]:
```suggestion
        :rtype: List[User]
```

File: interactions/api/models/guild.py [hunk 8]
```diff
@@ -1031,6 +1042,65 @@ async def delete_scheduled_event(self, event_id: int) -> None:
...
+        """
+
+        _role_id = role_id.id if isinstance(role_id, Role) else role_id
+        res = await self._client.modify_guild_role_position(
+            guild_id=int(self.id), position=position, role_id=_role_id, reason=reason
+        )
+        roles = [Role(**role, _client=self._client) for role in res]
+        return roles
+
+    async def get_bans(self) -> List[dict]:
```
Comment (i0bs) [inline interactions/api/models/guild.py:1093]:
```suggestion
    async def get_bans(self) -> List[User]:
```

File: interactions/api/models/guild.pyi [hunk 9]
```diff
@@ -265,6 +265,15 @@ class Guild(DictSerializerMixin):
...
     ) -> None: ...
+    async def get_all_channels(self) -> List[Channel]: ...
+    async def get_all_roles(self) -> List[Role]: ...
+    async def modify_role_position(
+        self,
+        role_id: Union[Role, int],
+        position: int,
+        reason: Optional[str] = None,
+    ) -> List[Role]: ...
+    async def get_bans(self) -> List[dict]: ...
```
Comment (i0bs) [inline interactions/api/models/guild.pyi:276]:
```suggestion
    async def get_bans(self) -> List[User]: ...
```

File: interactions/api/models/message.py [hunk 10]
```diff
@@ -423,6 +427,26 @@ async def reply(
...
         return Message(**res, _client=self._client)
 
+    async def pin(self) -> None:
+        """Pins the message to its channel"""
+
+        await self._client.pin_message(channel_id=int(self.channel_id), message_id=int(self.id))
+
+    async def unpin(self) -> None:
+        """Unpins the message from its channel"""
+        await self._client.unpin_message(channel_id=int(self.channel_id), message_id=int(self.id))
```
Comment (i0bs) [inline interactions/api/models/message.py:438]:
```suggestion

        await self._client.unpin_message(channel_id=int(self.channel_id), message_id=int(self.id))
```

**Review by i0bs — CHANGES_REQUESTED** [review]
I also see that you're importing specific models per iteration in a method, which can affect performance. (although it should be a very tiny amount because it'll recall upon loaded memory) Is there any reason why they're being done in the methods instead of an import in the file itself?

File: interactions/api/models/guild.py (hunk 6 above)
Comment (sillyElvy) [reply interactions/api/models/guild.py:1104]: I don't understand why I should change that

**Comment by sillyElvy** [comment]
> I also see that you're importing specific models per iteration in a method, which can affect performance. (although it should be a very tiny amount because it'll recall upon loaded memory) Is there any reason why they're being done in the methods instead of an import in the file itself?

Circular import

File: interactions/api/models/guild.py (hunk 6 above)
Comment (i0bs) [reply interactions/api/models/guild.py:1104]: Every other helper method you have returns a converted model object, but you don't do it for this one. It's silly to keep it inconsistent.

Comment (sillyElvy) [reply interactions/api/models/guild.py:1104]: it's a 
`{reason: user}`
dict
so why should I not include the reasons for the ban?

Comment (i0bs) [reply interactions/api/models/guild.py:1104]: Why not return a map then?

Comment (sillyElvy) [reply interactions/api/models/guild.py:1104]: how would I do that? I never worked with map

Comment (i0bs) [reply interactions/api/models/guild.py:1104]: Map is just another way to say dictionary, it's a different term for other programming languages. Just return a dictionary where the key and value are mapped respectively with a User object to reason relation.

Comment (sillyElvy) [reply interactions/api/models/guild.py:1104]: ok, how would I typehint that?

Comment (sillyElvy) [reply interactions/api/models/guild.py:1104]: I just checked again
It's
`[{"user": user, "reason": reason}]`
actually

**Review by FayeDel — APPROVED** [review]
Tested selects on this particular PR, and got no errors on the boot of this instance, and of sending selects. Looks good to me :)

**Pull request merged**